cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -82,62 +85,67 @@ impl Node<Option<cst::Policies>> {
+            .enumerate()
+            .map(|(count, node)| (ast::PolicyID::from_string(format!("policy{count}")), node)))
     }
 
     /// convert `cst::Policies` to `ast::PolicySet`
-    pub fn to_policyset(&self, errs: &mut ParseErrors) -> Option<ast::PolicySet> {
+    pub fn to_policyset(&self) -> Result<ast::PolicySet, ParseErrors> {
         let mut pset = ast::PolicySet::new();
-        let mut complete_set = true;
+        let mut all_errs: Vec<ParseErrors> = vec![];
         // Caution: `parser::parse_policyset_and_also_return_policy_text()`
         // depends on this function returning a policy set with `PolicyID`s as
         // generated by `with_generated_policyids()` to maintain an invariant.
         for (policy_id, policy) in self.with_generated_policyids()? {
             // policy may have convert error
-            match policy.to_policy_or_template(policy_id, errs) {
-                Some(Either::Right(template)) => {
+            match policy.to_policy_or_template(policy_id) {
+                Ok(Either::Right(template)) => {
                     if let Err(e) = pset.add_template(template) {
                         match e {
                             PolicySetError::Occupied { id } => {
-                                errs.push(self.to_ast_err(ToASTErrorKind::DuplicateTemplateId(id)))
+                                all_errs.push(ParseErrors::singleton(
+                                    self.to_ast_err(ToASTErrorKind::DuplicateTemplateId(id)),
+                                ))
                             }
                         };
-
-                        complete_set = false
                     }
                 }
-                Some(Either::Left(inline_policy)) => {
+                Ok(Either::Left(inline_policy)) => {
                     if let Err(e) = pset.add_static(inline_policy) {
                         match e {
                             PolicySetError::Occupied { id } => {
-                                errs.push(self.to_ast_err(ToASTErrorKind::DuplicatePolicyId(id)))
+                                all_errs.push(ParseErrors::singleton(
+                                    self.to_ast_err(ToASTErrorKind::DuplicatePolicyId(id)),
+                                ))
                             }
                         };
-
-                        complete_set = false
                     }
                 }
-                None => complete_set = false,
+                Err(errs) => {
+                    all_errs.push(errs);
+                }
             };
         }
 
         // fail on any error
-        if complete_set {
-            Some(pset)
-        } else {
-            None
+        match all_errs.split_first() {
+            None => Ok(pset),
+            Some((first, rest)) => {
+                let mut first = first.clone();
+                rest.iter()
+                    .for_each(|errs| first.extend(errs.iter().cloned()));
+                Err(first)
```

Comment:
maybe make this a helper function `.flatten()` with type `Vec<ParseErrors> -> ParseErrors`?